TS 3867 - Improved qsort

AZNova opened the pull request:
Updated qsort algo to a median of 3 qsort.

This qsort is called from the SSLContextSorage destructor during an SSL reconfigure.  The previous version could run into a worst case scenario (already sorted, or nearly sorted) and consume all of the stack space in the recursive calls to qsort.  This version uses a median of three by selecting the left-most, middle and right-most values and determining the median of those three values to use as the pivot point in the algorithm.  This should avoid the worst case by always choosing a value "somewhere in the middle".

jpeach commented:
I ran the regression tests on OS X with this patch, and saw some [memory leaks](http://apaste.info/z1O). the ```test_certlookup``` test (part of ```make check```) shows these.

AZNova commented:
I have updated this pull request with a new commit.  I ran all of the regression tests using 'make check', and all passed.  I have tested this with approx 10k certs and it worked fine.  I have also deployed this to one of our servers in our production environment with no issues.

SolidWallOfCode commented on lib/ts/Vec.h:
```diff
@@ -964,41 +965,71 @@ qsort_Vec(C *left, C *right, bool (*lt)(C, C))
+    median = *(right-2);        // and that's the median of the 3
+
+    // now partition, pivoting on the median value
+    // l ptr is +1 b/c we already put the lowest in there, ignore it for now
```
Not quite right - l (left) is not necessarily the lowest value in the span, but we do know !(median < left) which makes it valid for being in its position relative to the pivot (median). Same for right-1, !(right < median) so it's correctly grouped already. (This is about the comment, the code is correct)

SolidWallOfCode commented on lib/ts/Vec.h:
```diff
@@ -964,41 +965,71 @@ qsort_Vec(C *left, C *right, bool (*lt)(C, C))
+    C *l = left + 1, *r = right-2;
+
+    // move l and r until they have something to do
+    while (lt(median, *(r-1))) {
```
Note the original code doesn't check the bounds because we know the `lt` check will stop before `l < r` in any case (since !(median < median).

AZNova commented:
Updated the commit.  Moved the swap() above its initial usage.  Clarified comments.  Included amc's updated test_vec.cc which calls the new qsort as part of regression testing.
